### PR TITLE
Adding more world API methods

### LINKED ISSUE #5116: API to /locatebiome programmatically
**Is your feature request related to a problem? Please describe.**
Not necessarily a problem but a headache saver. 

**Describe the solution you'd like**
Basically just a way to locatebiome without needing NMS.

**Describe alternatives you've considered**
pain, agony even.

**Additional context**
Apparently, it's trivial.

### PR DESCRIPTION
**Pull request by Machine-Maker**
Adds api getters for [these world properties](https://minecraft.gamepedia.com/Custom#JSON_format).

Closes #5116

### CODE REVIEW

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
++
++    boolean isShrunk();
++
++    boolean hasSkylight();
++
++    boolean hasCeiling();
++
++    boolean isPiglinSafe();
++
++    boolean isBedWorks();
```
Comment (FranKusmiruk) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+    boolean doesBedWork();
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
++
++    boolean isRespawnAnchorWorks();
```
Comment (FranKusmiruk) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+    boolean doesRespawnAnchorWork();
```

**Review by FranKusmiruk — COMMENTED** [review]
Just wondering but are these fine without javadoc?

**Comment by Machine-Maker** [comment]
> Just wondering but are these fine without javadoc?

totally forgot... lol

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index 75c9225bd07b4c4b6dcd702c52294a7637eb6884..118394ac4956908571298d84afc75ef97e9b7e1e 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -3363,6 +3363,138 @@ public interface World extends PluginMessageRecipient, Metadatable {
+     @Nullable
+     public Location locateNearestStructure(@NotNull Location origin, @NotNull StructureType structureType, int radius, boolean findUnexplored);
+ 
++    // Paper start
++
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]: Drop this

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
++    /**
++     * Locates the nearest biome based on an origin, biome type, and radius to search
++     * Step defaults to 8
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Locates the nearest biome based on an origin, biome type, and radius to search.
+     * Step defaults to {@code 8}.
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
++    @Nullable
++    Location locateNearestBiome(@NotNull Location origin, @NotNull Biome biome, int radius, int step);
++
++    /**
++     * Checks if the world:
++     * evaporates water
++     * dries sponges
++     * has lava spread faster and further
++     *
++     * @return true if ultrawarm, false if not
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Checks if the world:
+     * <ul>
+     *     <li>evaporates water</li>
+     *     <li>dries sponges</li>
+     *     <li>has lava spread faster and further</li>
+     * </ul>
+     *
+     * @return whether the world is ultrawarm
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean isUltrawarm();
++
++    /**
++     * Checks if the world is natural
++     * If false, compasses will spin randomly in the world
++     * if true, nether portals will spawn zombified piglins
++     *
++     * @return true or false
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Checks if the world is natural.
+    * <p>
+     * If {@code false}, compasses will spin randomly in the world.
+     * If {@code true}, nether portals will spawn Zombified Piglins.
+     *
+     * @return whether the world is natural
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean isNatural();
++
++    /**
++     * Checks if the world is shrunk
++     * where 1 block is 8 blocks in the overworld
++     * like the nether
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Checks if the world is shrunk where 1 block is 8 blocks in the overworld like the Nether.
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
++     *
++     * @return true if shrunk
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether the world is shrunk
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean isShrunk();
++
++    /**
++     * Checks if the world has skylight access
++     *
++     * @return true if it has access
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether there is skylight
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean hasSkylight();
++
++    /**
++     * Checks if the world has a bedrock ceiling
++     *
++     * @return true if it has a bedrock ceiling
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether the world has a bedrock ceiling
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean hasCeiling();
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+    boolean hasBedrockCeiling();
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
++
++    /**
++     * Checks if piglins will turn into zombified piglins
++     * in this world
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Checks if Piglins will convert to Zombified Piglins in this world.
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
++     *
++     * @return true if piglins will NOT transform
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether Piglins will <i>not</i> transform
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean isPiglinSafe();
++
++    /**
++     * Checks if beds work
++     *
++     * @return true if beds work
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether beds work
```

What if they don't work?

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean doesBedWork();
++
++    /**
++     * Checks if respawn anchors work
++     *
++     * @return true if they work
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether respawn anchors work
```

What if they don't work?

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean doesRespawnAnchorWork();
++
++    /**
++     * Checks if this world supports raids
++     *
++     * @return true if supported
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether this world supports raids
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean hasRaids();
++
++    /**
++     * Checks if this world has a fixed time
++     *
++     * @return true if fixed
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * @return whether the world has fixed time
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
++     */
++    boolean hasCeiling();
++
++    /**
++     * Checks if piglins will turn into zombified piglins
++     * in this world
++     *
++     * @return true if piglins will NOT transform
++     */
++    boolean isPiglinSafe();
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]: This feels weird, as every other method is not inverse. Might be good to rename this?

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
++     *
++     * @return true if fixed
++     */
++    boolean isFixedTime();
++
++    /**
++     * Gets the collection of materials that burn
++     * infinitely in this world
++     *
++     * @return the infinite burn materials
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]:
```suggestion
+     * Gets the collection of materials that burn infinitely in this world.
+     *
+     * @return the materials that will forever stay lit by fire
```

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
++    @NotNull
++    Collection<Material> getInfiniburn();
++
++    /**
++     * Broadcast a message to all players
++     * in this world
++     *
++     * @param message Message to broadcast
++     */
++    void broadcastMessage(@NotNull String message);
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]: Components

File: Spigot-API-Patches/0218-More-World-API.patch [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
++
++    /**
++     * Broadcast a message to all players
++     * with a permission in a world
++     *
++     * @param message Message to broadcast
++     * @param permission Permission to check
++     */
++    void broadcastMessage(@NotNull String message, @NotNull String permission);
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-More-World-API.patch]: Components

File: Spigot-Server-Patches/0542-More-World-API.patch [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
++    }
++
++    @Override
++    public boolean isNatural() {
++        return getHandle().getDimensionManager().isNatural();
++    }
++
++    @Override
++    public boolean isShrunk() {
++        return getHandle().getDimensionManager().h();
```
Comment (Proximyst) [inline Spigot-Server-Patches/0542-More-World-API.patch]: OBFHELPER

File: Spigot-Server-Patches/0542-More-World-API.patch [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
++    }
++
++    @Override
++    public boolean isFixedTime() {
++        return getHandle().getDimensionManager().isFixedTime();
++    }
++
++    @Override
++    public Collection<org.bukkit.Material> getInfiniburn() {
++        return com.google.common.collect.Sets.newHashSet(com.google.common.collect.Iterators.transform(getHandle().getDimensionManager().q().getTagged().iterator(), CraftMagicNumbers::getMaterial));
```
Comment (Proximyst) [inline Spigot-Server-Patches/0542-More-World-API.patch]: OBFHELPER

**Review by Proximyst — DISMISSED** [review]

File: Spigot-API-Patches/0218-More-World-API.patch (hunk 18 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0218-More-World-API.patch]: I was just taking them directly as they were laid out in DimensionManager

File: Spigot-API-Patches/0236-More-World-API.patch [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
++    @NotNull
++    Collection<Material> getInfiniburn();
++
++    /**
++     * Broadcast a message to all players
++     * in this world
++     *
++     * @param message Message to broadcast
++     */
++    void broadcastMessage(@NotNull BaseComponent[] message);
```
Comment (kashike) [inline Spigot-API-Patches/0236-More-World-API.patch]: Remove this method from this pull request - tracking in #4830.

File: Spigot-API-Patches/0236-More-World-API.patch [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
++
++    /**
++     * Broadcast a message to all players
++     * with a permission in a world
++     *
++     * @param message Message to broadcast
++     * @param permission Permission to check
++     */
++    void broadcastMessage(@NotNull BaseComponent[] message, @NotNull String permission);
```
Comment (kashike) [inline Spigot-API-Patches/0236-More-World-API.patch]: Remove this method from this pull request - tracking in #4830.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0236-More-World-API.patch (hunk 25 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0236-More-World-API.patch]: Done

File: Spigot-API-Patches/0237-More-World-API.patch [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
++    @NotNull
++    Collection<Material> getInfiniburn();
++
++    /**
++     * Broadcast a message to all players
++     * in this world
++     *
++     * @param message Message to broadcast
++     */
++    void broadcastMessage(@NotNull String message);
```
Comment (kashike) [inline Spigot-API-Patches/0237-More-World-API.patch]: These two methods will end up deprecated as part of #4842 - is it worth even adding them now?

**Review by kashike — DISMISSED** [review]

File: Spigot-API-Patches/0237-More-World-API.patch (hunk 26 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0237-More-World-API.patch]: Yeah, you're right. I'll just leave a note to add the Component method after adventure get merged

**Comment by Machine-Maker** [comment]
Ok, I removed the broadcastMessage methods that were holding this PR up. Should be good to go now.

**Review by electronicboy — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Comment by Machine-Maker** [comment]
`Rebased after the NMS repackage`

File: Spigot-API-Patches/0283-More-World-API.patch [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
++     *
++     * @return true if ultrawarm, false if not
++     */
++    boolean isUltrawarm();
++
++    /**
++     * Checks if the world is natural.
++     * <p>
++     *     If {@code false}, compasses will spin randomly in the world.
++     *      * if {@code true}, nether portals will spawn zombified piglins.
```
Comment (MiniDigger) [inline Spigot-API-Patches/0283-More-World-API.patch]:
```suggestion
+     *      if {@code true}, nether portals will spawn zombified piglins.
```

**Review by MiniDigger — APPROVED** [review]
nitpick

**Review by kennytv — APPROVED** [review]

**Pull request merged**